### DISCUSSION TITLE
In LINQ what method syntax can do but query syntax cannot do about Index.

### DISCUSSION THREAD

**Discussion by chengyanbing** [dotnet/csharplang#5036] [Language Ideas]
The following example shows something that method syntax can do but query syntax cannot do in LINQ. 
Maybe I don't know how to do, if you know, please tell me.

static void Main(string[] args)
    {
        SearchText(@"D:\MyProgram\AspNetCore.Docs", "*.cshtml", "<link rel=\"icon\" href=\"data"); 
    }

    public static void SearchText(string searchPath, string searchFile, string searchText)
    {
        try
        {
            var filesAndLines = Directory.EnumerateFiles(searchPath, searchFile, SearchOption.AllDirectories)
                .SelectMany((file, fileID) =>
                    File.ReadLines(file).Select((line, lineID) => new { file, fileID, line, lineID = lineID + 1 }))
                .Where(allLines => allLines.line.Contains(searchText))
                .Select(allLines => new { File = allLines.file, Line = allLines.line, LineID = allLines.lineID });
            foreach (var f in filesAndLines)
            {
                Console.WriteLine($"{f.File}  -->  {f.Line}  ({f.LineID})");
            }
            Console.WriteLine($"{filesAndLines.Count()} lines found.");
        }
        catch (UnauthorizedAccessException uAEx)
        {
            Console.WriteLine(uAEx.Message);
        }
        catch (PathTooLongException pathEx)
        {
            Console.WriteLine(pathEx.Message);
        }
    }

Here I use it to search  _<link rel="icon" href="data_  because the windows explorer manager does dot support special characters like blank or slash or quote, etc. 
Method syntax can use argument **index**  in **Select** mothed or **SelectMany** mothed although this usage looks complicated, but query syntax cannot do that(Maybe I don't know how to do it, if you know, please tell me.). 
Can we implement it with a let statement? Maybe like this:

var filesAndLines = from file in Directory.EnumerateFiles(searchPath, searchFile, SearchOption.AllDirectories)
                                    let fileID = SetIndex(file)      // This is just my imagination
                                    from line in File.ReadLines(file)
                                    let lineID = SetIndex(line)     // This is  just my imagination
                                    where line.Contains(searchText)
                                    select new
                                    {
                                        File = file,
                                        Line = line,
                                        FileID = fileID,
                                        LineID = lineID
                                    };

**Comment by svick** [comment]
You can already get pretty close if you add an `Indexed` extension method:

```c#
public static IEnumerable<(T value, int index)> Indexed<T>(this IEnumerable<T> source) => source.Select((x, i) => (x, i));

from filePair in Directory.EnumerateFiles(searchPath, searchFile, SearchOption.AllDirectories).Indexed()
from linePair in File.ReadLines(filePair.value).Indexed()
where linePair.value.Contains(searchText)
select new
{
    File = filePair.value,
    Line = linePair.value,
    FileID = filePair.index,
    LineID = linePair.index
};
```

Though to make this truly nice, you would need tuple deconstruction in LINQ (https://github.com/dotnet/csharplang/issues/189). With that, the query would be:

```c#
from (file, fileID) in Directory.EnumerateFiles(searchPath, searchFile, SearchOption.AllDirectories).Indexed()
from (line, lineID) in File.ReadLines(file).Indexed()
where line.Contains(searchText)
select new
{
    File = file,
    Line = line,
    FileID = fileID,
    LineID = lineID
};
```

**Reply by chengyanbing** [reply]
Wa! That's great! I'll try.
